### ISSUE TITLE
Extend machineInfo to include NUMA topolgy

### ISSUE THREAD

**Issue by vikaschoudhary16** [google/cadvisor#1521]
Hello folks, I am working on a PoC to enable NUMA aware workloads in a Kubernetes deployment. I feel cadvisor should be enhanced to discover NUMA topology. Something similar to like libvirt does and one can see NUMA topology using `virsh capabilities`. This might be useful for any higher application interested in leveraging NUMA.
Thoughts?

**Comment by vikaschoudhary16** [comment]
cc @jeremyeder @timstclair @ConnorDoyle @derekwaynecarr

**Comment by ConnorDoyle** [comment]
cc @balajismaniam

**Comment by derekwaynecarr** [comment]
So the topology with cells and cpus listed?

Would you do this on MachineInfo?

Got a response format in mind?

**Comment by vikaschoudhary16** [comment]
@derekwaynecarr 

> So the topology with cells and cpus listed?

yes. But instead of 'cells', I would prefer 'numa_nodes' for better readability. 

> Would you do this on MachineInfo?

Yes.

> Got a response format in mind?

Does following format sounds reasonable:

```
type NUMANode struct {
	Id int `json:"numa_node_id"`
	// Per-numa-node memory
	Memory uint64  `json:"memory"`
	Cores  []Core  `json:"cores"`
}

type MachineInfo struct {
               .............
               .......
               // NUMA Topology
               // Describes cpu cores and memory available with each NUMA node
               NUMATopology   [ ] NUMANode `json:"numa_topolgy"`
               ........
               .........
}         

```

Thanks
-Vikas

**Comment by timothysc** [comment]
@vikaschoudhary16 If you have a POC in place.  Recommended to make a [WIP/POC] PR here to get feedback.

**Comment by vishh** [comment]
+1 to exposing numa info. NUMA support in k8s should be dealt with in a k8s proposal though.

**Comment by vikaschoudhary16** [comment]
@vishh thanks a lot and here is the discussion draft for k8s  proposal. PTAL
https://docs.google.com/document/d/19gj6XJSan3HiQ4tnbYBw6lxgbjASbtb-7sAtbDgZduA/edit

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
